keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,107 @@ always dispatched immediately.  Following is how the other requests
 are dispatched at a given apiserver.
 
 The concurrency limit of an apiserver is divided among the non-exempt
-priority levels in proportion to their assured concurrency shares.
-This produces the assured concurrency value (ACV) for each non-exempt
-priority level:
+priority levels, and higher ones can do a limited amount of borrowing
+from lower ones.
+
+Non-exempt priority levels are ordered in a total order for the
+purpose of borrowing currently-unused concurrency.  The ordering is
+based (first) on increasing value of a spec field whose name is
+`priority` and whose value is constrained to lie in the range 0
+through 10000 inclusive and (second) on increasing name of the
+PriorityLevelConfiguration object.  Priority levels that appear later
+in the order are considered to have lower priority.
+
+Borrowing is done on the basis of the current situation, with no
+consideration of opportunity cost and no pre-emption when the
+situation changes.
+
+An apiserver assigns three concurrency limits to each non-exempt
+priority level, with a constraint that means there are only two
+degrees of freedom.
+
+- The ***LendableConcurrencyLimit*** is the number of seats that are
+  statically (i.e., before any borrowing takes place) assigned to this
+  level and can be dynamically borrowed by higher levels.
+- The ***NonLendableConcurrencyLimit*** is the number of seats that
+  are statically assigned to this level and can _not_ be borrowed by
+  higher levels.
+- The ***NominalConcurrencyLimit*** is the number of seats statically
+  assigned to this level and is the sum of the
+  LendableConcurrencyLimit and the NonLendableConcurrencyLimit.
+
+Each non-exempt PriorityLevelConfiguration's spec has an
+`assuredConcurrencyShares`, which has existed since APF was introduced
+and may not be zero, and a `lendableConcurrencyShares` field, which is
+being added in the midst of the lifetime of the `v1beta2` version of
+the API and may be any value between zero and
+`assuredConcurrencyShares` inclusive (default is zero).  Each
+apiserver allocates NominalConcurrencyLimits in proportion to
+`assuredConcurrencyShares` and LendableConcurrncyLimit in
+corresponding propotion:
+
+```
+NominalConcurrencyLimit(i)  = ceil( SCL * assuredConcurrencyShares(i)  / sum_assured )
+LendableConcurrencyLimit(i) = ceil( SCL * lendableConcurrencyShares(i) / sum_assured )
+NonLendableConcurrencyLimit(i) = NominalConcurrencyLimit(i) - LendableConcurrencyLimit(i)
+sum_assured  = sum[priority levels k] assuredConcurrencyShares(k)
+```
+
+where SCL is the apiserver's concurrency limit.
+
+Borrowing is further limited by a practical consideration: we do not
```

Comment:
Let me float an alternative design.

Each PL (speaking of the in-memory structures only) is assigned:
* a `concurrency` number
* a `max_borrow` concurrency number (higher priorities may borrow more)
* a `max_lend` concurrency number (lower priorities may lend more)

Define some rate R which is the maximum speed at which PLs may borrow or lend concurrency.

In the dispatch loop for a PL, we compute a `drain_time` number. This should be, in real time units, how long we expect it to take to complete all the requests executing + enqueued. We also pre-compute, for each PL, an expected average `ideal_drain_time` assuming it is half loaded and has the default assigned concurrency.

So, if `drain_time` < `ideal_drain_time`, and a seat is free, and the PL has not reached the rate limit R, it will lend concurrency to the global pool. (which can be tracked via the sync.Atomic functions and not be as expensive as a mutex.)

if `drain_time` > `ideal_drain_time`, on the other hand, then we can attempt to borrow (also limited by R) to get an additional seat.

To ensure that the global pool will always be greater than zero, so that PLs which have lent can reclaim when needed, we define some global min. PLs which have currently borrowed must check that the pool has not fallen below the min, and if it has, return what they have borrowed rather than using it again to service another request.

PLs may not borrow or lend more than their configured maxes.

The design above is effectively (I think) a PID controller that's only using the P (proportional) term. It'd be easy and possibly useful to add in a D (derivative) term to be more responsive to rapidly changing load levels.

I can think of alternative conditions (other than drain_time) which may be superior, so this is both an idea about a basic borrowing/lending system, and an idea about the conditions under which PLs should borrow or lend.

Reply:
This is a fairly different design, with big details omitted.

Reply:
In slack, we've agreed to try this KEP design first and since the API would be equivalent, the above could be tried at a later date if we wanted.